File: codechef/PSHTRG/PSHTRG.cpp
#include <stdio.h>
#include "malloc.h"
#define N_max 100000
#define segment_length_max 45

int N, Q;
long long original_array[N_max];

struct segment {
	int number_of_elements_stored;
	long long elements_required_and_their_indices[segment_length_max][2];	
	int left, right;
	struct segment* left_child;
	struct segment* right_child;
};

int merge(long long first_array[][2], int number_of_elements_stored_in_first_array, long long second_array[][2], int number_of_elements_stored_in_second_array, long long merged_array[][2]) {	
	int c1 = 0, c2 = 0;
	int c = 0;

	while(c < segment_length_max && c < number_of_elements_stored_in_first_array + number_of_elements_stored_in_second_array) {	
		if(c1 >= number_of_elements_stored_in_first_array) {
			merged_array[c][0] = second_array[c2][0];
			merged_array[c][1] = second_array[c2][1];
			c2++;
		} else if(c2 >= number_of_elements_stored_in_second_array) {
			merged_array[c][0] = first_array[c1][0];
			merged_array[c][1] = first_array[c1][1];
			c1++;
		} else if(second_array[c2][0] > first_array[c1][0]) {
			merged_array[c][0] = second_array[c2][0];
			merged_array[c][1] = second_array[c2][1];
			c2++;
		} else {
			merged_array[c][0] = first_array[c1][0];
			merged_array[c][1] = first_array[c1][1];
			c1++;
		}
		c++;
	}

	return c;
}

void fill_node(int start, int end, struct segment* parent) {
	parent -> left = start;
	parent -> right = end;
	
	int i;	

	if(start == end) {
		parent -> number_of_elements_stored = 1;
		parent -> elements_required_and_their_indices[0][0] = original_array[start];
		parent -> elements_required_and_their_indices[0][1] = start;
		parent -> left_child = NULL;
		parent -> right_child = NULL;
		return;
	}
	
	int mid = (start + end)/2;
	parent -> left_child = (struct segment*) malloc(sizeof(struct segment));
	parent -> right_child = (struct segment*) malloc(sizeof(struct segment));
	fill_node(start, mid, parent -> left_child);
	fill_node(mid + 1, end, parent -> right_child);

	parent -> number_of_elements_stored = merge(parent -> left_child -> elements_required_and_their_indices, parent -> left_child -> number_of_elements_stored,
		parent -> right_child -> elements_required_and_their_indices, parent -> right_child -> number_of_elements_stored, parent -> elements_required_and_their_indices);
	
	return;
}

int get_sorted_array_in_range(int left_limit_of_range, int right_limit_of_range, struct segment* parent, long long sorted_array_in_range[][2]) {
	int i;	
	if(left_limit_of_range <= (parent -> left) && right_limit_of_range >= (parent -> right)) {
		for(i = 0; i < parent -> number_of_elements_stored; i++) {
			sorted_array_in_range[i][0] = parent -> elements_required_and_their_indices[i][0];
			sorted_array_in_range[i][1] = parent -> elements_required_and_their_indices[i][1];
		}
		return parent -> number_of_elements_stored;
	}
	
	long long left_array[segment_length_max][2];
	int left_array_len = 0;
	if(!(left_limit_of_range > (parent -> left_child -> right) || right_limit_of_range < (parent -> left_child -> left))) {
		left_array_len = get_sorted_array_in_range(left_limit_of_range, right_limit_of_range, parent -> left_child, left_array);
	}
		
	long long right_array[segment_length_max][2];
	int right_array_len = 0;
	if(!(left_limit_of_range > (parent -> right_child -> right) || right_limit_of_range < (parent -> right_child -> left))) {
		right_array_len = get_sorted_array_in_range(left_limit_of_range, right_limit_of_range, parent -> right_child, right_array);
	}
	
	if(left_array_len == 0) {
		for(i = 0; i < right_array_len; i++) {
			sorted_array_in_range[i][0] = right_array[i][0];
			sorted_array_in_range[i][1] = right_array[i][1];
		}
		return right_array_len;
	}
	
	if(right_array_len == 0) {
		for(i = 0; i < left_array_len; i++) {
			sorted_array_in_range[i][0] = left_array[i][0];
			sorted_array_in_range[i][1] = left_array[i][1];
		}
		return left_array_len;
	}
	
	int size_of_sorted_array_in_range = merge(left_array, left_array_len, right_array, right_array_len, sorted_array_in_range);	

	return size_of_sorted_array_in_range;
}

long long getMaximumPerimeter(int left_limit_of_range, int right_limit_of_range, struct segment* root_node) {
	long long sorted_array_in_range[segment_length_max][2];
	int size_of_sorted_array_in_range = get_sorted_array_in_range(left_limit_of_range, right_limit_of_range, root_node, sorted_array_in_range);
	int i;
	
	/*
	for(i = 0; i < size_of_sorted_array_in_range; i++) {
		printf("%lld %lld ", sorted_array_in_range[i][0], sorted_array_in_range[i][1]);
	}
	printf("\n");
	*/

	for(i = 0; i < size_of_sorted_array_in_range - 2; i++) {
		if(sorted_array_in_range[i][0] < sorted_array_in_range[i + 1][0] + sorted_array_in_range[i + 2][0])
			return sorted_array_in_range[i][0] + sorted_array_in_range[i + 1][0] + sorted_array_in_range[i + 2][0];
	}
	
	return 0;
}

	

void update_element(long long index_updated, int element, struct segment* parent) {
	int index_of_searched_element = segment_length_max;
	int i;		
	for(i = 0; i < parent -> number_of_elements_stored; i++) {
		if(parent -> elements_required_and_their_indices[i][1] == index_updated) {
			index_of_searched_element = i;
			break;
		}
	}
	
	long long temp[1][2];
	temp[0][0] = element;
	temp[0][1] = index_updated;

	if(index_of_searched_element != segment_length_max) {
		for(i = index_of_searched_element;i < parent -> number_of_elements_stored - 1; i++) {
			parent -> elements_required_and_their_indices[i][0] = parent -> elements_required_and_their_indices[i+1][0];
			parent -> elements_required_and_their_indices[i][1] = parent -> elements_required_and_their_indices[i+1][1];
		}
		parent -> number_of_elements_stored -= 1;
	}
	
	long long temp_array[segment_length_max][2];
	parent -> number_of_elements_stored = merge(temp, 1, parent -> elements_required_and_their_indices, parent -> number_of_elements_stored, temp_array);
	for(i = 0; i < parent -> number_of_elements_stored; i++) {
		parent -> elements_required_and_their_indices[i][0] = temp_array[i][0];
		parent -> elements_required_and_their_indices[i][1] = temp_array[i][1];
	}
	
	if(parent -> left_child != NULL) {
		if(index_updated <= parent -> left_child -> right && index_updated >= parent -> left_child -> left)
			update_element(index_updated, element, parent -> left_child);
	}

	if(parent -> right_child != NULL) {
		if(index_updated <= parent -> right_child -> right && index_updated >= parent -> right_child -> left)
			update_element(index_updated, element, parent -> right_child);
	}
}

int main() {
	scanf("%d", &N);
	scanf("%d", &Q);
	int i, q_type;
	int l, r;
	long long pos, val;

	for(i = 0; i < N; i++) {
		scanf("%lld", &original_array[i]);
	}

	struct segment* root_node = (struct segment*) malloc(sizeof(struct segment));
	fill_node(0, N-1, root_node);

	for(i = 0; i < Q; i++) {
		scanf("%d", &q_type);
		if(q_type == 2) {
			scanf("%d", &l);
			scanf("%d", &r);
			printf("%lld\n", getMaximumPerimeter(l - 1, r - 1, root_node));
		} else {
			scanf("%lld", &pos);
			scanf("%lld", &val);
			update_element(pos - 1, val, root_node);
		}
	}

	return 0;
}

File: codechef/PSHTRG/PSHTRG.java
/*
5 4
3 1 8 9 7
2 1 5
1 2 12
2 1 3
2 2 5
*/

import java.io.*;
import java.util.*;

class PSHTRG {
	static int N, Q;
	static ArrayList<Integer> original_array = new ArrayList<Integer>();
	static int magic_number = 45;
	
	static class Segment {
		ArrayList<ArrayList<Integer>> elements_required_and_their_indices = new ArrayList<ArrayList<Integer>>();
		int left;
		int right;
		Segment left_child;
		Segment right_child;
	}

	public static ArrayList<ArrayList<Integer>> merge(ArrayList<ArrayList<Integer>> first_array, ArrayList<ArrayList<Integer>> second_array) {
		ArrayList<ArrayList<Integer>> merged_array = new ArrayList<ArrayList<Integer>>();	
		int c1 = 0, c2 = 0;
		int size_of_first_array = first_array.size();
		int size_of_second_array = second_array.size();
		int c = 0;

		while(c < magic_number && c < size_of_first_array + size_of_second_array) {
			merged_array.add(new ArrayList<Integer>());	
			if(c1 >= size_of_first_array) {
				merged_array.get(c).add(second_array.get(c2).get(0));
				merged_array.get(c).add(second_array.get(c2).get(1));
				c2++;
			} else if(c2 >= size_of_second_array) {
				merged_array.get(c).add(first_array.get(c1).get(0));
				merged_array.get(c).add(first_array.get(c1).get(1));
				c1++;
			} else if(second_array.get(c2).get(0) > first_array.get(c1).get(0)) {
				merged_array.get(c).add(second_array.get(c2).get(0));
				merged_array.get(c).add(second_array.get(c2).get(1));
				c2++;
			} else {
				merged_array.get(c).add(first_array.get(c1).get(0));
				merged_array.get(c).add(first_array.get(c1).get(1));
				c1++;
			}
			c++;
		}

		return merged_array;
	}

	public static ArrayList<ArrayList<Integer>> fill_node(int start, int end, Segment parent) {
		parent.left = start;
		parent.right = end;
		
		if(start == end) {
			ArrayList<Integer> element_to_be_added_and_its_index = new ArrayList<Integer>();
			element_to_be_added_and_its_index.add(original_array.get(start));
			element_to_be_added_and_its_index.add(start);
			parent.elements_required_and_their_indices.add(element_to_be_added_and_its_index);
			parent.left_child = null;
			parent.right_child = null;

			return parent.elements_required_and_their_indices;
		}
		
		int mid = (start + end)/2;
		parent.left_child = new Segment();
		parent.right_child = new Segment();
		ArrayList<ArrayList<Integer>> left_array = fill_node(start, mid, parent.left_child);
		ArrayList<ArrayList<Integer>> right_array = fill_node(mid + 1, end, parent.right_child);

		parent.elements_required_and_their_indices = merge(left_array, right_array);

		return parent.elements_required_and_their_indices;
	}

	public static Long getMaximumPerimeter(int left_limit_of_range, int right_limit_of_range, Segment root_node) {
		ArrayList<ArrayList<Integer>> sorted_array_in_range = get_sorted_array_in_range(left_limit_of_range, right_limit_of_range, root_node);
		for(int i = 0; i < sorted_array_in_range.size() - 2; i++) {
			if(sorted_array_in_range.get(i).get(0) < sorted_array_in_range.get(i + 1).get(0) + sorted_array_in_range.get(i + 2).get(0))
				return new Long(sorted_array_in_range.get(i).get(0)) + new Long(sorted_array_in_range.get(i + 1).get(0)) + new Long(sorted_array_in_range.get(i + 2).get(0));
		}
		
		return 0L;
	}

	public static ArrayList<ArrayList<Integer>> get_sorted_array_in_range(int left_limit_of_range, int right_limit_of_range, Segment parent) {
		if(left_limit_of_range <= parent.left && right_limit_of_range >= parent.right) {
			return parent.elements_required_and_their_indices;
		}
		
		ArrayList<ArrayList<Integer>> left_array = new ArrayList<ArrayList<Integer>>();
		if(!(left_limit_of_range > parent.left_child.right || right_limit_of_range < parent.left_child.left)) {
			left_array = get_sorted_array_in_range(left_limit_of_range, right_limit_of_range, parent.left_child);
		}
		
		ArrayList<ArrayList<Integer>> right_array = new ArrayList<ArrayList<Integer>>();
		if(!(left_limit_of_range > parent.right_child.right || right_limit_of_range < parent.right_child.left)) {
			right_array = get_sorted_array_in_range(left_limit_of_range, right_limit_of_range, parent.right_child);
		}

		if(left_array.isEmpty())
			return right_array;
		
		if(right_array.isEmpty())
			return left_array;
		
		return merge(left_array, right_array);
	}
	
	public static void update_element(int index_updated, int element, Segment parent) {
		int index_of_searched_element = magic_number;		
		for(int i = 0; i < parent.elements_required_and_their_indices.size(); i++) {
			if(parent.elements_required_and_their_indices.get(i).get(1) == index_updated) {
				index_of_searched_element = i;
				break;
			}
		}
		
		ArrayList<ArrayList<Integer>> temp = new ArrayList<ArrayList<Integer>>();
		ArrayList<Integer> element_to_be_added_and_its_index = new ArrayList<Integer>();
		element_to_be_added_and_its_index.add(element);
		element_to_be_added_and_its_index.add(index_updated);
		temp.add(element_to_be_added_and_its_index);

		if(index_of_searched_element == magic_number) {
			parent.elements_required_and_their_indices = merge(temp, parent.elements_required_and_their_indices);
		} else {
			parent.elements_required_and_their_indices.remove(index_of_searched_element);
			parent.elements_required_and_their_indices = merge(temp, parent.elements_required_and_their_indices);
		}
		
		if(parent.left_child != null) {
			if(index_updated <= parent.left_child.right && index_updated >= parent.left_child.left)
				update_element(index_updated, element, parent.left_child);
		}

		if(parent.right_child != null) {
			if(index_updated <= parent.right_child.right && index_updated >= parent.right_child.left)
				update_element(index_updated, element, parent.right_child);
		}
	}

	public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		Q = sc.nextInt();
		int i;

		for(i = 0; i < N; i++) {
			original_array.add(sc.nextInt());
		}

		Segment root_node = new Segment();
		fill_node(0, N-1, root_node);

		for(i = 0; i < Q; i++) {
			if(sc.nextInt() == 2) {
				System.out.println(getMaximumPerimeter(sc.nextInt() - 1, sc.nextInt() - 1, root_node));
			} else {
				update_element(sc.nextInt() - 1, sc.nextInt(), root_node);
			}
		}
	}
}
